google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py

```diff
@@ -168,17 +176,35 @@ def ack(self, items: Sequence[requests.AckRequest]) -> None:
 
         # We must potentially split the request into multiple smaller requests
         # to avoid the server-side max request size limit.
-        ack_ids = (item.ack_id for item in items)
+        items_gen = iter(items)
+        ack_ids_gen = (item.ack_id for item in items)
         total_chunks = int(math.ceil(len(items) / _ACK_IDS_BATCH_SIZE))
 
         for _ in range(total_chunks):
-            request = gapic_types.StreamingPullRequest(
-                ack_ids=itertools.islice(ack_ids, _ACK_IDS_BATCH_SIZE)
-            )
-            self._manager.send(request)
-
-        # Remove the message from lease management.
-        self.drop(items)
+            future_reqs_dict = {req.ack_id: req for req in itertools.islice(items_gen, _ACK_IDS_BATCH_SIZE) if req.future}
+            requests_completed, requests_to_retry =  self._manager.send_unary_ack(
+                ack_ids=list(itertools.islice(ack_ids_gen, _ACK_IDS_BATCH_SIZE)),
+                future_reqs_dict=future_reqs_dict)
+
+            # Remove the completed messages from lease management.
+            self.drop(requests_completed)
+
+            # retry acks
+            retry_delay_gen = exponential_sleep_generator(initial=_MIN_ACK_MODACK_RETRY_DURATION_SECS,
```

Comment:
This seems to be happening with or without exactly once? This should happen only for exactly once subs. Am I understanding this correctly?

Reply:
This code path is only executed if there's futures with errors, and that only happens if 1) the user invokes a ack/modack_with_future method 2) the server comes back with an error for that ack id. Both of these only occur when the user is using exactly-once. Note we don't check if exactly-once is enabled at the stream level for this behavior. I think it's wise not to b/c we'll better handle edge cases around enablement/disablement. For example, if the server sends an ack error and we had just learned that exactly-once has been disabled at the stream level, we still want to tell the user the ack failed.

Reply:
I meant ack/modack_with_result, not *with_future.